Specify output directory with CLI

xmclark opened the pull request:
This PR addresses #235 

There is a new command line flag for the build command: `--out-dir` and `-d`. Supply a relative path and put your wasm somewhere else! Default is `pkg`. 

---

Make sure these boxes are checked! 📦✅

- [x] You have the latest version of `rustfmt` installed and have your 
      cloned directory set to nightly
```bash
$ rustup override set nightly
$ rustup component add rustfmt-preview --toolchain nightly
```
- [x] You ran `rustfmt` on the code base before submitting
- [x] You reference which issue is being closed in the PR text

✨✨ 😄 Thanks so much for contributing to wasm-pack! 😄 ✨✨

xmclark commented:
I would mark this WIP, until there is some feedback 😄

technetos commented on tests/all/manifest.rs:
```diff
@@ -157,6 +157,28 @@ fn it_creates_a_pkg_json_with_correct_files_on_node() {
+    let out_dir = PathBuf::from("../ferris/is/awesome");
+    let fixture = fixture::fixture("tests/fixtures/js-hello-world");
+    let step = wasm_pack::progressbar::Step::new(1);
+    wasm_pack::command::utils::create_pkg_dir(&fixture.path, &out_dir, &step).unwrap();
```
Can we space these out more, its awfully difficult to read.

fitzgen commented on tests/all/manifest.rs:
```diff
@@ -55,7 +55,7 @@ fn it_recognizes_a_map_during_depcheck() {
     let fixture = fixture::fixture(".");
     let step = wasm_pack::progressbar::Step::new(1);
-    wasm_pack::command::utils::create_pkg_dir(&fixture.path, &step).unwrap();
+    wasm_pack::command::utils::create_pkg_dir(&fixture.path, &PathBuf::from("pkg"), &step).unwrap();
```
Nitpick: this can be `Path::new("pkg")`, which won't allocate like how creating a temporary `PathBuf` will.

Also, since we are always using the default relative out dir in these tests, it makes sense to add a helper test utility that always uses "pkg" to `tests/all/utils/manifest.rs` and use that here.

fitzgen commented on src/command/build.rs:
```diff
@@ -21,6 +21,7 @@ pub(crate) struct Build {
     pub debug: bool,
     // build_config: Option<BuildConfig>,
     pub crate_name: String,
+    pub out_dir: String,
```
Nitpick: this should be a `PathBuf`

fitzgen commented on src/command/build.rs:
```diff
@@ -62,6 +62,9 @@ pub struct BuildOptions {
     // build_config: Option<BuildConfig>,
+
+    #[structopt(long = "out-dir", short = "d", default_value = "pkg")]
+    pub out_dir: String,
```
This should be a `PathBuf` and then we should use `structopt(parse(from_os_str))` just like the `path` member up above.

fitzgen commented on tests/all/manifest.rs:
```diff
@@ -157,6 +157,28 @@ fn it_creates_a_pkg_json_with_correct_files_on_node() {
 
+#[test]
+fn it_creates_a_pkg_json_in_out_dir() {
+    let out_dir = PathBuf::from("../ferris/is/awesome");
```
For tests, we should probably use either

* a custom out directory that is still within the crate, eg "out" instead of "pkg", or
* a temporary directory with `tempfile::TempDir`.

What we don't want is to create random directories that might conflict with other tests and might not get cleaned up at the end of the tests.

xmclark replied:
Well said. I will do a custom directory in the crate.

xmclark commented:
This solution still has a few issues and I am working through them.

data-pup commented:
++, thanks for taking care of this @xmclark :smile_cat:

xmclark commented:
@mgattozzi hi! It is pretty much done. I was able to test locally. I did have some questIons that I had posted in the issue, but if it is better to keep conversation here, I am happy to discuss it. 

We use the term “pkg” to rerference the generates directory in a couple different places. Now that there is a custom out dir, that is not totally accurate. I was wondering if anybody had input on this.

data-pup commented:
Tested this locally, ran into these two errors:

```
---- readme::it_copies_a_readme_default_path stdout ----
wasm-pack: copying test fixture '/home/user/Github/wasm-pack' to temporary directory '/tmp/.tmpUqvE1a/wasm-pack'
thread 'readme::it_copies_a_readme_default_path' panicked at 'should create pkg directory OK: Os { code: 17, kind: AlreadyExists, message: "File exists" }', libcore/result.rs:983:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- readme::it_creates_a_package_json_provided_path stdout ----
wasm-pack: copying test fixture '/home/user/Github/wasm-pack/tests/fixtures/js-hello-world' to temporary directory '/tmp/.tmpohMt6R/wasm-pack'
thread 'readme::it_creates_a_package_json_provided_path' panicked at 'should create pkg directory OK: Os { code: 17, kind: AlreadyExists, message: "File exists" }', libcore/result.rs:983:5


failures:
    readme::it_copies_a_readme_default_path
    readme::it_creates_a_package_json_provided_path

test result: FAILED. 15 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out
```

This seems like a good start though! I think once these are fixed, and the branch is rebased to pass CI, this should be good to go.

xmclark commented:
I did the rebase and am having some issues getting the wasm-pack to finish building my local test app. I am getting output that suggests I am getting stuck part way through wasm-packing building a project. See my comments on this [PR](https://github.com/rustwasm/wasm-pack/pull/244/files#r212832458).

xmclark commented:
After the underscore/hyphens issue was resolved, I rebased and squashed again. I tested using a local test app and it works as expected. The canonical path is printed.

fitzgen commented:
It... looks like the travis ci vm ran out of disk space? https://travis-ci.org/rustwasm/wasm-pack/jobs/421425086#L809

```
---- manifest::it_creates_a_package_json_with_correct_keys_when_types_are_skipped stdout ----
wasm-pack: copying test fixture '/home/travis/build/rustwasm/wasm-pack' to temporary directory '/tmp/.tmpRkw5Nz/wasm-pack'
thread 'manifest::it_creates_a_package_json_with_correct_keys_when_types_are_skipped' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 28, kind: Other, message: "No space left on device" })', libcore/result.rs:983:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:221
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:477
   5: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:391
   6: rust_begin_unwind
             at libstd/panicking.rs:326
   7: core::panicking::panic_fmt
             at libcore/panicking.rs:77
   8: core::result::unwrap_failed
             at /checkout/src/libcore/macros.rs:26
   9: <core::result::Result<T, E>>::unwrap
             at /checkout/src/libcore/result.rs:782
  10: all::manifest::it_creates_a_package_json_with_correct_keys_when_types_are_skipped
             at tests/all/manifest.rs:200
  11: all::__test::TESTS::{{closure}}
             at tests/all/manifest.rs:196
  12: core::ops::function::FnOnce::call_once
             at /checkout/src/libcore/ops/function.rs:238
  13: <F as alloc::boxed::FnBox<A>>::call_box
             at libtest/lib.rs:1460
             at /checkout/src/libcore/ops/function.rs:238
             at /checkout/src/liballoc/boxed.rs:644
  14: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:103
```

I'll retrigger the job and see if that was a one off failure...

xmclark commented:
Looks like @fitzgen re-ran the job because the timestamps match up (thank you!). When I first saw this failure, I thought it was very bizarre and assumed it was a one-off. Now that it has been repeated...I have to figure out why...oh boy.

xmclark commented:
@fitzgen 
- [x] Try disabling cache: cargo in .travis.yml? 

I think this was the issue. Still not sure why.

I tested locally, and squashed all of the new commits. Did a local test too.

![image](https://user-images.githubusercontent.com/1364747/44959748-9574d200-aea8-11e8-8461-4e3a9b651dcf.png)
